contract_sale_invoicing/models/contract.py

```diff
@@ -9,23 +9,122 @@ class ContractContract(models.Model):
              '{} 23:59:59'.format(self.recurring_next_date)),
-        ])
-        if sales:
-            invoice_ids = sales.action_invoice_create()
-            invoices |= self.env['account.invoice'].browse(invoice_ids)[:1]
+            ]
+        if self.filter_with == 'analytic_account':
+            domain.append(('analytic_account_id', '=', self.group_id.id))
+        elif self.filter_with == 'contract':
+            domain.append(('contract_id', '=', self.id))
+        return domain
+
+    @api.multi
+    def _recurring_create_invoice(self, date_ref=False):
+        invoices = super()._recurring_create_invoice(date_ref)
+        for rec in self:
+            if not rec.invoicing_sales:
+                return invoices
+            so_domain = rec.get_sale_order_domain()
+            sales = self.env['sale.order'].search(so_domain)
+            if sales and rec.group_by == 'sale_order':
+                invoice_ids = sales.action_invoice_create()
+                invoices |= self.env['account.invoice'].browse(invoice_ids)[:1]
+
+    @api.multi
+    def _prepare_invoice_line_dict(self, contract_line_rec, invoice_line,
+                                   remain_qty):
+        return {
+            'invoice_id': False,
+            'uom_id': contract_line_rec.uom_id.id,
+            'product_id': invoice_line.get('product_id'),
+            'quantity': remain_qty or 0,
+            'discount': contract_line_rec.discount,
+            'contract_line_id': contract_line_rec.id,
+            'name': contract_line_rec.name,
+            'account_analytic_id': False,
+            'price_unit': contract_line_rec.price_unit
+        }
+
+    @api.multi
+    def _prepare_recurring_invoices_values(self, date_ref=False):
+        invoices_values = super()._prepare_recurring_invoices_values()
+        updated_invoices_values = []
+        for invoice_val in invoices_values:
+            invoice_line_values = {}
+            invoice_line_list = []
+            for invoice_line in invoice_val.get('invoice_line_ids', []):
+                invoice_line = invoice_line[2] or {}
+                contract_line_rec = self.env['contract.line'].\
+                    browse(invoice_line.get('contract_line_id', False))
+                if contract_line_rec and contract_line_rec.contract_id and\
+                        contract_line_rec.contract_id.invoicing_sales and \
+                        contract_line_rec.contract_id.group_by == 'contract':
+                    so_domain = \
+                        contract_line_rec.contract_id.get_sale_order_domain()
+                    order_ids = self.env['sale.order'].search(so_domain)
+                    sale_order_line_product_qty = {}
+                    for order_id in order_ids:
+                        if not order_id.order_line.mapped('invoice_lines'):
+                            for line in order_id.order_line:
+                                if sale_order_line_product_qty.\
```

Comment:
@sbejaoui In our scenario we need to merge them because the pricelist on the invoice applies a different price based on the quantity. If @nikul-serpentcs adds a boolean field on the contract to merge sale order lines with the same product, would that work for you?

Reply:
There isn't a pricelist on invoices. The standard behavior is to keep the price when creation an invoice from SO.

Reply:
If you merge lines, how will compute the invoiced qty in SO?

Reply:
@sbejaoui Good point. 

@nikul-serpentcs Let's not merge and keep one invoice line per sale order line.

Reply:
@nikul-serpentcs , you resolved this discussion but you still merging invoice lines